src/parse/context.rs

```diff
@@ -18,12 +22,101 @@ pub trait Context {
     {
         item.parse(self)
     }
+
+    /// Add a [SimpleExtensionUri] to this context. Must return an error for duplicate
+    /// anchors or when the URI is not supported.
+    ///
+    /// This function must eagerly resolve and parse the simple extension, returning an
+    /// error if either fails.
+    fn add_simple_extension_uri(
+        &mut self,
+        simple_extension_uri: &SimpleExtensionUri,
+    ) -> Result<&SimpleExtensions, ContextError>;
+
+    /// Returns the simple extensions for the given simple extension anchor.
+    fn simple_extensions(
+        &self,
+        anchor: &Anchor<SimpleExtensionUri>,
+    ) -> Result<&SimpleExtensions, ContextError>;
+}
+
+/// Parse context errors.
+#[derive(Debug, Error, PartialEq)]
+pub enum ContextError {
+    /// Undefined reference to simple extension.
+    #[error("undefined reference to simple extension with anchor `{0}`")]
+    UndefinedSimpleExtension(Anchor<SimpleExtensionUri>),
+
+    /// Duplicate anchor for simple extension.
+    #[error("duplicate anchor `{0}` for simple extension")]
+    DuplicateSimpleExtension(Anchor<SimpleExtensionUri>),
+
+    /// Unsupported simple extension URI.
+    #[error("unsupported simple extension URI: {0}")]
+    UnsupportedURI(String),
 }
 
 #[cfg(test)]
 pub(crate) mod tests {
-    #[derive(Default)]
-    pub struct Context {}
+    use std::collections::{hash_map::Entry, HashMap};
 
-    impl super::Context for Context {}
+    use crate::parse::{
+        context::ContextError, proto::extensions::SimpleExtensionUri,
+        text::simple_extensions::SimpleExtensions, Anchor,
+    };
+
+    /// A test context.
+    ///
+    /// This currently mocks support for simple extensions (always returns an
+    /// empty simple extensions).
+    pub struct Context {
+        empty_simple_extensions: SimpleExtensions,
+        simple_extensions: HashMap<Anchor<SimpleExtensionUri>, SimpleExtensionUri>,
+    }
+
+    impl Default for Context {
+        fn default() -> Self {
+            Self {
+                empty_simple_extensions: SimpleExtensions {},
+                simple_extensions: Default::default(),
+            }
+        }
+    }
+
+    impl super::Context for Context {
+        fn add_simple_extension_uri(
+            &mut self,
+            simple_extension_uri: &crate::parse::proto::extensions::SimpleExtensionUri,
+        ) -> Result<&SimpleExtensions, ContextError> {
+            match self.simple_extensions.entry(simple_extension_uri.anchor()) {
+                Entry::Occupied(_) => Err(ContextError::DuplicateSimpleExtension(
+                    simple_extension_uri.anchor(),
+                )),
+                Entry::Vacant(entry) => {
+                    // This is where we would resolve and then parse.
+                    // This check shows the use of the unsupported uri error.
+                    if let "http" | "https" | "file" = simple_extension_uri.uri().scheme() {
```

Comment:
Some dialects do not require or expect a scheme (I'll get examples when I return from vacation).  It may be useful to make this configurable.

Reply:
That should be documented here: https://github.com/substrait-io/substrait/blob/e564cf0b3b769491c19016d92d195b5f62361775/proto/substrait/extensions/extensions.proto#L18-L20

Please note that this is just a test context, and not the actual context that I want to eventually provide in this crate. In that case yes, the resolver should be configurable.

Enjoy your vacation!